hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyRequest.java

```diff
@@ -775,10 +775,19 @@ protected OmKeyInfo prepareFileInfo(
       dbKeyInfo.setReplicationConfig(replicationConfig);
 
       // Construct a new metadata map from KeyArgs.
-      // Clear the old one when the key is overwritten.
-      dbKeyInfo.getMetadata().clear();
-      dbKeyInfo.getMetadata().putAll(KeyValueUtil.getFromProtobuf(
-          keyArgs.getMetadataList()));
+      // Clear the old one when the key is overwritten unless it is a key overwrite
+      // using optimistic commit.
+      if (!keyArgs.hasOverwriteUpdateID()) {
+        dbKeyInfo.getMetadata().clear();
+        dbKeyInfo.getMetadata().putAll(KeyValueUtil.getFromProtobuf(
+            keyArgs.getMetadataList()));
+      }
+      if (keyArgs.hasOverwriteObjectID()) {
+        dbKeyInfo.setOverwriteObjectID(keyArgs.getOverwriteObjectID());
+      }
+      if (keyArgs.hasOverwriteUpdateID()) {
+        dbKeyInfo.setOverwriteUpdateID(keyArgs.getOverwriteUpdateID());
+      }
```

Comment:
I'm not sure I understand this section in the original code either. It looks like we are copying the key metadata as it exists at the time of create, when it seems we should actually be doing this at the time of commit. Wherever this logic is happening, this is probably where we want to copy the native ACLS for the overwrite case as well.

Reply:
I will check about the ACLs. However I am pretty sure the ACLs and all other meta data is automatically copied from the original key, as the "openKey" entry it take from the original closedKey entry. I will double check on that.

Reply:
I was also not sure why the meta data is explicitly cleared here in the original code, while other key attributes are not. The metadata here, is a map of key value pairs that can be attached to a key. For the overwrite case, we just want to keep whatever is there hence the change here.

Reply:
The original code is confusing because it copies the existing key and then modifies into the new key that gets aded to the open key table, while leaving the original in the DB. I checked the fields being modified in the original and it looks correct, including preserving the ACLs.